Assets/Character/Scripts/PlayerState/PlayerState.cs

```diff
@@ -61,10 +70,30 @@ protected void EnableInputHandling(bool b)
 	protected void ChangeState(PlayerStateEnum newState)
 	{
         _playerInput = null;
-        _selfTransition = ((int)newState == AnimatorCommon.GetState(_animator));
-        AnimatorCommon.SetState(_animator, (int)newState);
+        int cost = Constants.GetStaminaCost(newState);
+
+        if (cost <= _stamina.getRemaining())
+        {
+            _selfTransition = ((int)newState == AnimatorCommon.GetState(_animator));
+            AnimatorCommon.SetState(_animator, (int)newState);
+        }
+        else
+        {
+            _selfTransition = false;
+            AnimatorCommon.SetState(_animator, (int)PlayerStateEnum.TestIdle);
+        }
+        _stamina.subtract(cost); // always remove stamina, even if won't transition
 	}
 
+    protected void RechargeStamina()

```

Comment:
private, unless we're implying substates can recharge differently. But i also mentioned above we might want stamina to recharge itself.